### PR TITLE
Parse zipped pdbs

### PR DESCRIPTION
**Pull request by atravitz**
Tips
* Comment "pre-commit.ci autofix" to have pre-commit.ci atomically format your PR.
  Since this will create a commit, it is best to make this comment when you are finished with your work.

Checklist
* [ ] Added a ``news`` entry

## Developers certificate of origin
- [ ] I certify that this contribution is covered by the MIT License [here](https://github.com/OpenFreeEnergy/openfe/blob/main/LICENSE) and the **Developer Certificate of Origin** at <https://developercertificate.org/>.

### CODE REVIEW

File: gufe/tests/data/181l.cif.gz
Comment (mikemhenry) [inline gufe/tests/data/181l.cif.gz:1]: a little confusing that we are not using a `.pdb.gz` file (looks like we are using `181l.cif` instead of `181l.pdb`?).

File: gufe/tests/test_proteincomponent.py [hunk 1]
```diff
@@ -120,9 +120,12 @@ def test_from_pdb_file(self, in_pdb_path):
         assert isinstance(p, ProteinComponent)
         assert p.name == "Steve"
 
-    def test_from_pdbx_file(self, PDBx_181L_path):
+    def test_from_pdbx_file(self, PDBx_181L_path, PDBx_181L_gz_path):
         p = self.cls.from_pdbx_file(str(PDBx_181L_path), name="Steve")
-
+        p_gz = self.cls.from_pdbx_file(str(PDBx_181L_gz_path), name="Steve")
+        # footgun? this test requires that the .pdb and .pdb.gz files contain the same information
```
Comment (mikemhenry) [inline gufe/tests/test_proteincomponent.py]: Yes, we should just gzip this as part of the test

File: gufe/tests/conftest.py [hunk 2]
```diff
@@ -153,6 +153,12 @@ def PDBx_181L_path():
         yield str(f)
 
 
+@pytest.fixture
+def PDBx_181L_gz_path():
+    with importlib.resources.path("gufe.tests.data", "181l.cif.gz") as f:
+        yield str(f)
+
+
```
Comment (mikemhenry) [inline gufe/tests/conftest.py]:
```suggestion
@pytest.fixture(scope="session")
def PDBx_181L_gz_path(PDBx_181L_path, tmp_path_factory):
    gz_path = tmp_path_factory.getbasetemp() / "181l.cif.gz"
    with open(PDBx_181L_path, "rb") as f_in, gzip.open(gz_path, "wb") as f_out:
        f_out.write(f_in.read())
    yield str(gz_path)
```

File: gufe/tests/conftest.py
Comment (mikemhenry) [inline gufe/tests/conftest.py]:
```suggestion
@pytest.fixture(scope="session")
```

Comment (mikemhenry) [reply gufe/tests/conftest.py]: Need both of these fixtures to have the same scope, since we are just resolving a data file path, session is safe.

File: gufe/tests/conftest.py (hunk 2 above)
Comment (mikemhenry) [reply gufe/tests/conftest.py]: we need to use a `tmp_path_factory` so our xdist workers don't trip over themselves

File: gufe/tests/test_proteincomponent.py (hunk 1 above)
Comment (mikemhenry) [reply gufe/tests/test_proteincomponent.py]: I've updated the fixture to handle this if you want to use that instead

**Review by mikemhenry — CHANGES_REQUESTED** [review]
I've got some ideas -- I'll take over the PR and have you review it

**Comment by mikemhenry** [comment]
I used ChatGPT 5.4 Pro with the context of `open_text_stream` and asked it to `add doc strings in the numpy doc format for open_text_stream. be clear but not overly verboise. be careful and don't make a mistake.`

File: gufe/utils.py [hunk 3]
```diff
@@ -1,10 +1,160 @@
...
+        header = bytes(header)
+
+        if raw.seekable():
+            raw.seek(0)
+            source = raw
+            source_is_caller_owned = raw_is_caller_owned
+        else:
+            source = io.BytesIO(header + raw.read())
+            source_is_caller_owned = False
+            stack.callback(source.close)
```
Comment (atravitz) [inline gufe/utils.py:155]: checking my understanding: this chunk is resetting us to the top so that header bytes aren't lost?

File: gufe/utils.py [hunk 4]
```diff
@@ -1,10 +1,160 @@
...
+
+        opener = _detect_opener(header)
+
+        stream: TextIO
+        if opener is not None:
+            stream = opener(source, "rt")
+            stack.callback(stream.close)
+        else:
```
Comment (atravitz) [inline gufe/utils.py:163]: checking my understanding: here we haven't found any magic signatures in the header, so we're assuming it's plain text, and will throw errors accordingly if someone passes in an unsupported compressed format?

File: gufe/tests/test_utils.py [hunk 5]
```diff
@@ -97,3 +100,132 @@ def test_ensure_filelike_default_mode():
...
+        with open_text_stream(plain_text_path) as f:
+            assert f.read() == PLAIN_TEXT.decode()
+
+    def test_compressed_path(self, compressed_path):
+        """All three compression formats are detected via magic bytes."""
+        with open_text_stream(compressed_path) as f:
+            assert f.read() == PLAIN_TEXT.decode()
+
+    def test_magic_bytes_not_extension(self, tmp_path):
+        """Compression is detected from magic bytes, not file extension."""
```
Comment (atravitz) [inline gufe/tests/test_utils.py:155]: 🎉

File: gufe/tests/test_utils.py [hunk 6]
```diff
@@ -97,3 +100,132 @@ def test_ensure_filelike_default_mode():
...
+            data = f.read()
+        with open_text_stream(_NonSeekableBytesIO(data)) as stream:
+            assert stream.read() == PLAIN_TEXT.decode()
+
+    def test_caller_stream_not_closed(self, plain_text_path):
+        """When a stream is passed in, it is not closed on context manager exit."""
+        with open(plain_text_path, "rb") as binary_stream:
+            with open_text_stream(binary_stream):
+                pass
+            assert not binary_stream.closed
```
Comment (atravitz) [inline gufe/tests/test_utils.py:184]: is this what you were able to resolve using `ExitStack`?

**Review by atravitz — COMMENTED** [review]
looks great, just some questions for my own understanding. thanks for picking this up to make it so robust!

File: gufe/utils.py (hunk 3 above)
Comment (mikemhenry) [reply gufe/utils.py:155]: Yes, if we can seek the stream we do BUT some streams are not seekable so then we concatenate what we have already read with the rest of the stream. The reason we don't just always do the concatenation is that then we have to load the entire stream/file into memory all the time. With this setup we can take advantage of the memory savings of seekable streams.

File: gufe/utils.py (hunk 4 above)
Comment (mikemhenry) [reply gufe/utils.py:163]: Yes, basically we assume that the binary data passed in can be decoded as UTF-8. I added this branch since we had some tests (I think it was this one https://github.com/OpenFreeEnergy/gufe/blob/main/gufe/tests/test_proteincomponent.py#L36) that needed to support that behavior. At one point I had this raise an error that was like "hey don't do that, this file can be opened as `r`" but after thinking about it, I didn't want to make things a bit more strict and risk breaking user code.

File: gufe/tests/test_utils.py (hunk 6 above)
Comment (mikemhenry) [reply gufe/tests/test_utils.py:184]: Yes and this case https://github.com/OpenFreeEnergy/gufe/pull/777/changes#diff-3147c4d7e8e168cbbcd2e7820f9c784fab1408923d8ee28dbd1ae6c7adb1592aR217 where a user might do something like
```
with open(data, "rb") as data_stream:
    with open_text_stream(data_stream) as decompressed_stream:
        for line in decompressed_stream:
            ...
```
and then we would end up in the case of a double close on `data_stream` as the context managers started closing things

File: gufe/utils.py [hunk 7]
```diff
@@ -1,10 +1,160 @@
...
+            raw_is_caller_owned = True
+
+        header = raw.read(_HEADER_READ_SIZE)
+        if not isinstance(header, (bytes, bytearray, memoryview)):
+            raise TypeError(
+                "Binary streams passed to open_text_stream must return bytes. "
+                "Text streams should be passed as text and will be yielded unchanged."
+            )
+        header = bytes(header)
+
```
Comment (mikemhenry) [inline gufe/utils.py:140]: Add comment  The reason we don't just always do the concatenation is that then we have to load the entire stream/file into memory all the time. With this setup we can take advantage of the memory savings of seekable streams.

File: gufe/utils.py [hunk 8]
```diff
@@ -111,15 +111,17 @@ def open_text_stream(
...
         return
 
     with ExitStack() as stack:
+        raw: BinaryIO
         if isinstance(path_or_stream, (str, PathLike)):
             raw = stack.enter_context(open(path_or_stream, "rb"))
             raw_is_caller_owned = False
         else:
-            raw = path_or_stream
+            raw = cast(BinaryIO, path_or_stream)
```
Comment (mikemhenry) [inline gufe/utils.py:130]: add comment on cast calls explaing that this helps mypy but doesn't do any runtime stuff

File: gufe/utils.py [hunk 9]
```diff
@@ -1,10 +1,160 @@
...
+    """Detach a wrapper without closing the caller-owned binary stream."""
+    try:
+        stream.detach()
+    except ValueError:
+        # Already closed/detached.
+        pass
+
+
+@contextmanager
+def open_text_stream(
```
Comment (mikemhenry) [inline gufe/utils.py]:
```suggestion
def magic_open(
```

**Review by mikemhenry — APPROVED** [review]
image
